lib/AFP/AbstractFilePath/Internal/Common.hs

```diff
@@ -0,0 +1,692 @@
+  where
+    f x | isWindows = dropTrailingPathSeparator $ BS.map toLower $ normalise x
+        | otherwise = dropTrailingPathSeparator $ normalise x
+
+
+isRelative :: ShortByteString -> Bool
+isRelative x = BS.null drive || isRelativeDrive drive
+    where drive = takeDrive x
+
+{- c:foo -}
+-- From [1]: "If a file name begins with only a disk designator but not the
+-- backslash after the colon, it is interpreted as a relative path to the
+-- current directory on the drive with the specified letter."
+isRelativeDrive :: ShortByteString -> Bool
+isRelativeDrive x =
+    maybe False (not . hasTrailingPathSeparator . fst) (readDriveLetter x)
+
+
+isAbsolute :: ShortByteString -> Bool
+isAbsolute = not .isRelative
+
+
+-- Information for validity functions on Windows. See [1].
+isBadCharacter :: Word -> Bool
+isBadCharacter x = x >= _nul && x <= 31
+  || x `elem`
+      [ _less
+      , _greater
+      , _colon
+      , _quotedbl
+      , _bar
+      , _question
+      , _asterisk
+      ]
+
+badElements :: [ShortByteString]
+badElements = fmap fromString
+    ["CON","PRN","AUX","NUL","CLOCK$"
+    ,"COM1","COM2","COM3","COM4","COM5","COM6","COM7","COM8","COM9"
+    ,"LPT1","LPT2","LPT3","LPT4","LPT5","LPT6","LPT7","LPT8","LPT9"]
+
+isValid :: ShortByteString -> Bool
+isValid path
+  | BS.null path = False
+  | _nul `BS.elem` path = False
+  | isPosix = True
+  | otherwise =
+      not (BS.any isBadCharacter x2) &&
+      not (any f $ splitDirectories x2) &&
+      not (isJust (readDriveShare x1) && BS.all isPathSeparator x1) &&
+      not (isJust (readDriveUNC x1) && not (hasTrailingPathSeparator x1))
+    where
+      (x1,x2) = splitDrive path
+      f x = BS.map toUpper (BS.dropWhileEnd isSpace $ dropExtensions x) `elem` badElements
+
+
+makeValid :: ShortByteString -> ShortByteString
+makeValid path
+  | BS.null path = BS.singleton _underscore
+  | isPosix = BS.map (\x -> if x == _nul then _underscore else x) path
+  | isJust (readDriveShare drv) && BS.all isPathSeparator drv = BS.take 2 drv `BS.append` (fromString "drive")
+  | isJust (readDriveUNC drv) && not (hasTrailingPathSeparator drv) =
+      makeValid (drv `BS.append` (BS.singleton pathSeparator) `BS.append` pth)
+  | otherwise = joinDrive drv $ validElements $ validChars pth
+
+  where
+    (drv,pth) = splitDrive path
+
+    validChars = BS.map f
+    f x = if isBadCharacter x then _underscore else x
+
+    validElements = joinPath . map g . splitPath
+    g x = h a `BS.append` b
+        where (a,b) = BS.break isPathSeparator x
+    h x = if BS.map toUpper (BS.dropWhileEnd isSpace a) `elem` badElements then (BS.snoc _underscore a) <.> b else x
+        where (a,b) = splitExtensions x
+
+
+isFileName :: ShortByteString -> Bool
+isFileName filepath =
+  not (foldr (\a b -> BS.singleton a `BS.isInfixOf` filepath || b) False pathSeparators) &&
+  not (BS.null filepath) &&
+  not (_nul `BS.elem` filepath)
+
+
+hasParentDir :: ShortByteString -> Bool
+hasParentDir filepath =
```

Comment:
This one doesn't exist in `filepath`, but in `path` and I found it worthwhile: https://github.com/commercialhaskell/path/blob/d24c3407819b728b8430c6fc17a9904967c9da88/src/Path/Internal/Include.hs#L80